lua/weapons/weapon_spraypaint/cl_init.lua

```diff
@@ -1,66 +1,576 @@
+end
+
+function SWEP:DoParticle(pos, endpos, color, size)
+    for i = 1, 5 do
+        local matrix, opos, oang = self:DrawWorldModel(true)
+        pos = opos + oang:Up() * 5
 
-        if (tr.StartPos:Distance(tr.HitPos) < 125) then
-            surface.DrawCircle(ScrW() / 2, ScrH() / 2, 11, Color(255, 255, 255, 8))
-            surface.DrawCircle(ScrW() / 2, ScrH() / 2, 10, Color(0, 0, 0, 10))
+        if (self:GetOwner() == LocalPlayer() and not self:GetOwner():ShouldDrawLocalPlayer()) then
+            local vm = self:GetOwner():GetViewModel()
+            local bone = vm:LookupBone("plunger")
+
+            if (bone) then
+                local matrix = vm:GetBoneMatrix(bone)
+
+                if (matrix) then
+                    pos = matrix:GetTranslation() + matrix:GetAngles():Forward() * -5 + matrix:GetAngles():Up() * -1
+                end
+            end
         end
+
+        local dir = (endpos - pos):GetNormal()
+        local tv = dir * 500
+        pos = pos + dir * math.Rand(0, tv:Length()) * FrameTime()
+        self.SprayEmitter = self.SprayEmitter or ParticleEmitter(pos)
+        if (not SpraypaintParticleEmitter) then return end
+        self.SprayEmitter:SetPos(pos)
+        local particle = self.SprayEmitter:Add(string.format("particle/smokesprites_00%02d", math.random(7, 16)), pos)
+        --if DisableConvar:GetBool() then return end
+        particle:SetColor(color.r, color.g, color.b, color.a)
+        particle:SetStartAlpha(color.a)
+        particle:SetVelocity(tv)
+        particle:SetGravity(Vector(0, 0, 0))
+        particle:SetLifeTime(0)
+        particle:SetLighting(false)
+        particle:SetDieTime(math.Rand(0.1, 0.3))
+        particle:SetStartSize(0)
+        particle:SetEndSize(size * 0.25)
+        particle:SetEndAlpha(0)
+        particle:SetCollide(true)
+        particle:SetBounce(0)
+        particle:SetRoll(math.Rand(0, 360))
+        particle:SetRollDelta(0.01 * math.Rand(-40, 40))
+        particle:SetAirResistance(100)
     end
+end
+
+if CLIENT then
+    surface.CreateFont("spraypaint_ammocounter", {
+        font = "Coolvetica", --  Use the font-name which is shown to you by your operating system Font Viewer, not the file name
+        extended = false,
+        size = 128,
+        weight = 500,
+        blursize = 2,
+        scanlines = 0,
+        antialias = true,
+        underline = false,
+        italic = false,
+        strikeout = false,
+        symbol = false,
+        rotary = false,
+        shadow = false,
+        additive = false,
+        outline = false,
+    })
+
+    hook.Add("PreDrawEffects", "DrawSprayPaintHUD", function()
+        local wep = LocalPlayer():GetActiveWeapon()
+        if (not IsValid(wep) or wep:GetClass() ~= "weapon_spraypaint") then return end
+        local trace = wep:GetTrace()
+        if (not trace.Hit or trace.HitPos:Distance(EyePos()) > wep:GetPaintDistance()) then return end
+        if (trace.HitSky) then return end
```

Comment:
it would make sense to put this with the previous if statement (the set of trace-related failures to cancel drawing...)